### DISCUSSION TITLE
API Identify

### DISCUSSION THREAD

**Discussion by james-rae** [fgpv-vpgf/fgpv-vpgf#4116] [Team Posts]
This is mostly near-final draft of [Identify API](https://github.com/fgpv-vpgf/fgpv-vpgf/issues/3545) spec , but there is an item that can be discussed.

## General Behavior and Guts

Function will trigger a map identify on a map point anywhere (regardless if on or off screen). The current map scale will be used in the math to calculate the tolerance of the identified area.

Class `Map` in the API gets method `identify`. Single parameter of type `geo.XY | geo.XYLiteral`. No return value. Method will have very nice jsdoc header.

Guts of method will translate the parameter to a native map point (esri point), create a fake click event (essentially an object with property `mapPoint` set to the native map point), and call `identify` method in the viewers Identify Service. Done thanks. 

### Banning MouseEvent as Param Type

The Map API click observable tosses up a `MouseEvent`, which is the likely source of point information to handle the “re-click” use case. However, this event also contains properties about where on the screen the click happened. This opens up the potential for an implementer to feed in untrue data to the API (e.g. grab and save MouseEvent, user pans map, more code then uses the saved value to trigger identify. The API click observable tells lies about where on the screen the new click happened).

Luckily the workaround to the ban is simple, and intuitive thanks to the nice jsdoc spec that will be written.

```text
var myMouseEvent = [api map click observable result];
// various deeds occur
// time to re-click
mapApi.identify(myMouseEvent.xy);
```

### Chatty API

The Identify Service triggers a number of API observables when it runs. These include:

- Map click
- A feature stream on that click as results return
- Layer Group identify (if we have `Query` identify mode turned on)

In particular, the map click one is interesting. A physical click didn’t happen, and the fake click is known to the host page. Should we be reporting it back? 

- A 3rd party plugin may not be aware of click fakery, and needs to detect the fake click.
- Conversely, a 3rd party plugin may only want to react to real mouse events, and thus should not detect the fake click.
- Given we have no screen information, a plugin can technically inspect the click event contents and equate the lack of screen co-ords with fake clickery. This knowledge would not be obvious to Jimmy McRandomDeveloper.
- An API option to control if fake clicks trigger the click observable?
- The click observable also contains the feature stream on it, which can be a useful tool for the caller if they want to process the identify results. They can also use the identify observable in the Layer Group, but it takes a bit more code to make it act like a stream (an array of promises need to be dealt with as they resolve).
- Is this being over-thunk?

## Stretch Goals

In Identify Service, locate `// TODO: include the actual referenced layer` and resolve the hardcoded string it refers to.

In geoApi, add `supportsIdentify` boolean property to layer records. Have wms, dynamic, and feature records return `true`. 
Then can adjust this code in Identify Service

```text
.filter(layerRecord => typeof layerRecord.identify !== 'undefined')
// becomes
.filter(layerRecord => layerRecord.supportsIdentify)
```

**Comment by AleksueiR** [comment]
I'm a bit confused about banning MouseEvent. You are referring to a case where `myMouseEvent.xy` and `myMouseEvent.screenXY` are maliciously modified to not indicate the same point (a screen point matching a point on the map). Is this right? 

I think we should report back real clicks as well as fake clicks and let the host page sort them out. To help with that, we can add a `fake-click` flag to the event (the flag should really be called something like `non-user-initiated-event` to make it sound fancy).

**Comment by james-rae** [comment]
The concern was more for accidental use, not malicious.  E.g. 

```
var myMouseEvent = [api map click observable result];
// various deeds occur
// user pans the map

// plugin 1 decides it's time to re-click
mapApi.identify(myMouseEvent); // map point is no longer aligned with screen point

// plugin 2 hears click observable, consumes mis-aligned data 
// 3rd party developer is mildly confused by, or possibly unaware of funny state.
```

I'm fine with adding an extra property (would suggest `api-initiated` or similar).
Both extra property approach, and the ban approach (or manually nulling of screen properties on an api initiated click observable) will require some documentation on why this is, and what situations a developer needs to be concerned with. 

Real impact is only if a 3rd party developer is concerned with screen co-ordinates. Banning/nulling makes it impossible to use mistaken co-ords, but also prevents a clever dev from writing fancy code to mitigate as much as possible.  Having the event flag means the dev needs to be aware of the flag and the risks it identifies, otherwise they can easily go about consuming incorrect data.

**Comment by mweech** [comment]
Having a flag that will identify as api-initiated sounds like a good strategy to me, especially if we accompany the API with the necessary documentation.  

At the end of the day, we are only pointing out that the plugin developer should not assume there has been no subsequent map interaction since the last identify occured, and therefore non-geo based coordinates are unreliable.  That's pretty straight forward to understand.  

I think its best we provide context and don't artificially hide stuff as a general rule.

**Comment by milespetrov** [comment]
I think a click event should only be triggered when an actual click occurs. I suggest keeping the current map click stream as is for mouse clicks (and deprecating the returned feature stream as of `v4`) in favor of a new observable on the map instance named `identified`. This observable emits with the identify coordinates and the stream of feature results. We can include the `MouseEvent` object as well if the identify was triggered by a mouse click, otherwise it is `null`.

**Comment by james-rae** [comment]
Discussed Miles' suggestion at the Jamboree, decided the flag approach + outstanding documentation allows us the most flexibility in letting an implementer decide how they want to handle various incarnations of clicky-thingss

**Comment by james-rae** [comment]
Follow up note. Turns out WMS identify needs the screen co-ords, but is also smart enough to accept co-ords that are outside the current viewing window and return correctly.

Enhanced our identify routine so if it sees an identify request coming from the API (i.e. just map XY, no pixel co-ords), it will reverse engineer the appropriate screen coords (using ESRI API's tool for this) and use those.
